bgpd/bgp_fsm.c

```diff
@@ -1145,6 +1145,8 @@ int bgp_stop(struct peer *peer)
 
 		/* Reset peer synctime */
 		peer->synctime = 0;
+
+		bgp_bfd_deregister_peer(peer);
```

Comment:
is this needed ONLY when the state is Established?

Reply:
Yes, BFD session is created only after BGP is in Established state.

So once BGP enters Established ----> any other state, this is required.

Reply:
BFD will be usually deleted only when BGP configuration is removed not when BGP goes down. Rightly so bgp_bfd_deregister_peer is been called by peer_delete. If the parameters are changed then it is admin configuration that should trigger updating BFD parameters or BGP should only then deregister and register it. This may impact implimeantion where once BGP goes down BGP will wait for BFD to come back up again.

Reply:
> BFD will be usually deleted only when BGP configuration is removed not when BGP goes down. Rightly so bgp_bfd_deregister_peer is been called by peer_delete. If the parameters are changed then it is admin configuration that should trigger updating BFD parameters or BGP should only then deregister and register it. This may impact implimeantion where once BGP goes down BGP will wait for BFD to come back up again.

BFD session is created when BGP session is Established not just when BFD is configured in BGP. There is no point of having BFD session when BGP session in not Established.
Deleting BFD session when BGP session is DOWN was the base behaviour which was modified for a optimisation which led to side effect of BFD session having stale local address in some scenarios. This code change just intends to revert to old behaviour.

Reply:
Should we not use bgp_bfd_update_peer this function to update source address.

Reply:
> > BFD will be usually deleted only when BGP configuration is removed not when BGP goes down. Rightly so bgp_bfd_deregister_peer is been called by peer_delete. If the parameters are changed then it is admin configuration that should trigger updating BFD parameters or BGP should only then deregister and register it. This may impact implimeantion where once BGP goes down BGP will wait for BFD to come back up again.
> 
> BFD session is created when BGP session is Established not just when BFD is configured in BGP. There is no point of having BFD session when BGP session in not Established.
> Deleting BFD session when BGP session is DOWN was the base behaviour which was modified for a optimisation which led to side effect of BFD session having stale local address in some scenarios. This code change just intends to revert to old behaviour.

Old code could have been a workaround (not sure). BFD session is not removed administratively so why delete it. If parameters are changed then appropriate API can be called IMHO. As you pointed out it was done for optimisation as everytime BFD flaps it is removed and recreated again.